distributed/worker_state_machine.py

```diff
@@ -2077,109 +2096,146 @@ def _transition_resumed_released(
+            else:
+                ts.state = "resumed"
+                ts.next = "fetch"
+                return {}, []
 
     def _transition_cancelled_waiting(
         self, ts: TaskState, *, stimulus_id: str
     ) -> RecsInstrs:
-        if ts.done:
-            return {ts: "released"}, []
-        elif ts.previous in ("executing", "long-running"):
-            ts.state = ts.previous
-            return {}, []
+        """
+        See also
+        --------
+        _transition_cancelled_fetch
+        _transition_resumed_fetch
+        _transition_resumed_waiting
+        """
+        if ts.previous in ("executing", "long-running"):
+            if ts.done:
+                # execute just completed for a cancelled task.
+                # Discard output and possibly forget
+                return {ts: "released"}, []
+            else:
+                # Forget the task was cancelled to begin with
+                ts.state = ts.previous
+                return {}, []
         else:
             assert ts.previous == "flight"
-            ts.state = "resumed"
-            ts.next = "waiting"
-            return {}, []
+            if ts.done:
+                return {ts: "waiting"}, []
+            else:
+                ts.state = "resumed"
+                ts.next = "waiting"
+                return {}, []
 
     def _transition_cancelled_forgotten(
         self, ts: TaskState, *, stimulus_id: str
     ) -> RecsInstrs:
-        ts.next = "forgotten"
-        if not ts.done:
-            return {}, []
         return {ts: "released"}, []
 
     def _transition_cancelled_released(
         self, ts: TaskState, *, stimulus_id: str
     ) -> RecsInstrs:
         if not ts.done:
             return {}, []
-        self.executing.discard(ts)
-        self.long_running.discard(ts)
-        self.in_flight_tasks.discard(ts)
 
-        self._release_resources(ts)
-        return self._transition_generic_released(ts, stimulus_id=stimulus_id)
+        if ts.previous in ("executing", "long-running"):
+            # Calls _ensure_computing
+            recs, instr = self._transition_from_executing(ts)
+        else:
+            assert ts.previous == "flight"
+            # No need to call _ensure_communicating, since it's done at the end of
+            # every transition loop
+            recs, instr = {}, []
+
+        return merge_recs_instructions(
+            (recs, instr),
+            self._transition_generic_released(ts, stimulus_id=stimulus_id),
+        )
 
     def _transition_executing_released(
         self, ts: TaskState, *, stimulus_id: str
     ) -> RecsInstrs:
-        ts.previous = ts.state
+        """We can't stop executing a task just because the scheduler asked us to,
+        so we're entering cancelled state and waiting until it completes.
+        """
+        assert ts.state in ("executing", "long-running")
+        ts.previous = cast(Literal["executing", "long-running"], ts.state)
```

Comment:
I'm curious, what does this literal cast give us?

Reply:
ts.state is a Literal[<all states>]
ts.previous is a Literal[executing, long-running, flight, None]

mypy is not smart enough to figure out the assertion on the line above guarantees the intersection of the two domains.